### PR TITLE
Prevent duplicated leaves in the backend

### PR DESCRIPTION
**Pull request by davxy**
I've noticed that if, as fork choice rule, we don't use the "longest chain" then we may end up inserting duplicate entries in the backend leaves set.

The reason is basically this check:

https://github.com/paritytech/substrate/blob/986fab5ab303e566f637bb57e3c9c9b7df1d369e/client/db/src/lib.rs#L1348-L1349

Here we're using `number <= best_num` as a shortcut to detect if a block can't be a duplicate. That is, if `number > best_num` then we assume that the block can't be a duplicate and thus we don't check the second condition. But this is correct only if we are using the longest chain rule.

This bug is always triggered on parachains built with Cumulus when we change the state of an already imported block (with number greater that the current best) to be the new best (via `import_block_as_new_best` [here](https://github.com/paritytech/cumulus/blob/007ec5b796d6ffcc0fae1fb99264705243df0c0f/client/consensus/common/src/parachain_consensus.rs#L355)).

The effect is that the code within [this](https://github.com/paritytech/substrate/blob/986fab5ab303e566f637bb57e3c9c9b7df1d369e/client/db/src/lib.rs#L1508) scope is executed and the block re-added to the leaves list (as a  duplicate).

In general, this is not a big issue since the duplicates are removed as soon as a block with greater number is finalized, but can be annoying if we need to know how many leaves we have at a given point in time (and I need it :-) )

---

This PR fix the issue by removing the "shortcut"

---

**⚠⚠⚠ ACHTUNG ⚠⚠⚠**

I'm not 100% familiar with some of the code not-executed after this modification (i.e. within [this](https://github.com/paritytech/substrate/blob/986fab5ab303e566f637bb57e3c9c9b7df1d369e/client/db/src/lib.rs#L1508) scope).

 **Please double check that there are no unwanted side effects**

### CODE REVIEW

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -1345,8 +1345,7 @@ impl<Block: BlockT> Backend<Block> {
 
 			let parent_hash = *pending_block.header.parent_hash();
 			let number = *pending_block.header.number();
-			let existing_header =
-				number <= best_num && self.blockchain.header(BlockId::hash(hash))?.is_some();
+			let existing_header = self.blockchain.header(BlockId::hash(hash))?.is_some();
```
Comment (cheme) [inline client/db/src/lib.rs:1348]: test at line 1561 becomes redundant.

**Review by cheme — APPROVED** [review]
Change looks good to me. Would probably wait that original author confirm it was just an optimization. But looks like it to me.

File: client/db/src/lib.rs (hunk 1 above)
Comment (davxy) [reply client/db/src/lib.rs:1348]: Do you mean this?

https://github.com/paritytech/substrate/blob/b8706b4e28709631387b74a2ce846f5fa71e3127/client/db/src/lib.rs#L1561-L1563

If yes, here we're testing the `parent_hash` and not `hash`

Comment (cheme) [reply client/db/src/lib.rs:1348]: Oh sure you're right.

Comment (bkchr) [inline client/db/src/lib.rs:1348]: We could keep the check if we combine it with this here: https://github.com/paritytech/substrate/pull/11716/files#diff-d631f4b5aa83ad9f2088c7d13236ad51b6a4ffdb870927f071532d9904bc2c5fR264

Aka that we not compare against `best_num` and instead compare against `highest_leaf`.

Comment (arkpar) [reply client/db/src/lib.rs:1348]: We could check if parent is a known leaf instead.

**Comment by arkpar** [comment]
This was an optimization indeed. Querying the database for a missing header is a relatively expensive operation, as missing queries are not cached. However it only "works" when the first block for the new top number is imported. Better implementation would be to cache all known block hashes for a few recent tree heights. For now I think it is OK to remove this  check or to replace it with something as simple. As suggested in the [comments](https://github.com/paritytech/substrate/pull/11941#discussion_r934506723) above

**Comment by davxy** [comment]
@bkchr @arkpar done

**Review by arkpar — APPROVED** [review]

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -3534,7 +3541,24 @@ pub(crate) mod tests {
 			header.hash()
 		};
 
-		let block3_fork = insert_header_no_head(&backend, 3, block2, Default::default());
+		let block3_fork = {
+			let mut op = backend.begin_operation().unwrap();
+			backend.begin_state_operation(&mut op, BlockId::Number(1)).unwrap();
```
Comment (bkchr) [inline client/db/src/lib.rs:3546]:
```suggestion
			backend.begin_state_operation(&mut op, BlockId::Hash(block2)).unwrap();
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**